### PR TITLE
Python Error Diagnostics. Bug fix for incrementing range generation

### PR DESCRIPTION
**Pull request by amccaskey**
Fixed a bug in incrementing ranges like the following
```python
for i in range(0, 4, 2):
  ...
```

Setup error diagnostics for AST compilation to MLIR. 

New error diagnostics for invalid code. Just a few examples to get an idea for what it looks like
```python 
@cudaq.kernel()
def ex0():
    l = [] # can't create an empty list 
```
```bash
cudaq.kernel.ast_bridge.CompilerError: tmp.py:11: error: CUDA Quantum does not support allocating lists of zero size (no dynamic memory allocation)
         (offending source -> l = [])
```
```python
@cudaq.kernel
def test(i : int, f:float):
    cudaq.dbg.ast.print_i64(i)
    q = cudaq.qvector(2)
    ry(f, q[f]) # qubit extraction with a float instead of int
```
```bash
cudaq.kernel.ast_bridge.CompilerError: tmp.py:11: error: invalid index variable type used for qvector extraction (f64)
         (offending source -> q[f])
```

### CODE REVIEW

File: python/tests/kernel/test_observe_kernel.py [hunk 1]
```diff
@@ -136,7 +136,7 @@ def kernel(theta: float, phi: float):
     assert len(energies) == 50
 
     @cudaq.kernel
-    def kernel(thetas: list):
+    def kernel(thetas: list[float]):
```
Comment (khalatepradnya) [inline python/tests/kernel/test_observe_kernel.py:139]: This test is failing on Python v3.8, we can skip `list[float]` on it and have another test using `List[float]`

File: docs/sphinx/install.rst [hunk 2]
```diff
@@ -521,7 +521,7 @@ To get started with DGX Cloud, you can
 `request access here <https://www.nvidia.com/en-us/data-center/dgx-cloud/trial/>`__.
 Once you have access, `sign in <https://ngc.nvidia.com/signin>`__ to your account,
 and `generate an API key <https://ngc.nvidia.com/setup/api-key>`__. 
-`Install the NGC CLI <https://ngc.nvidia.com/setup/installers/cli>`__
+`Install the NGC CLI <https://ngc.nvidia.com/setup/cli>`__
```
Comment (khalatepradnya) [inline docs/sphinx/install.rst:524]: The old link is working now. I see that this new one is giving error...

Comment (bmhowe23) [reply docs/sphinx/install.rst:524]: Lol, thanks for letting me know. I'll close this one.

**Comment by bmhowe23** [comment]
Oops, sorry about that! Wrong PR. I re-opened this one.

File: python/cudaq/kernel/ast_bridge.py [hunk 3]
```diff
@@ -414,8 +537,8 @@ def visit_FunctionDef(self, node):
 
             arguments = node.args.args
             if len(arguments):
-                raise RuntimeError(
-                    "inner function defs cannot have arguments yet.")
+                self.emitFatalError(
+                    "inner function defintions cannot have arguments.", node)
```
Comment (khalatepradnya) [inline python/cudaq/kernel/ast_bridge.py:541]: Typo - `definitions`

Comment (amccaskey) [reply python/cudaq/kernel/ast_bridge.py]: thanks, fixed

File: python/cudaq/kernel/ast_bridge.py [hunk 4]
```diff
@@ -2461,8 +2645,7 @@ def compile_to_mlir(astModule, **kwargs):
     try:
         pm.run(bridge.module)
     except:
-        print(bridge.module)
-        raise RuntimeError("could not canonicalize code for '{}'.".format(
+        raise RuntimeError("could not compile code for '{}'.".format(
             bridge.name))
```
Comment (khalatepradnya) [inline python/cudaq/kernel/ast_bridge.py:2649]: Refactoring opportunity - consistent wording across error messages, e.g. capitalization and tool name. Can be done later.

**Comment by khalatepradnya** [comment]
Q: Is it possible to add a failing test that demonstrates the new diagnostic and error messaging capability?

**Review by khalatepradnya — APPROVED** [review]
Looks good to me.

**Review by bmhowe23 — APPROVED** [review]

**Comment by khalatepradnya** [comment]
Addresses error handling in issue https://github.com/NVIDIA/cuda-quantum/issues/1251

**Pull request merged**